test/extensions/filters/udp/udp_proxy/udp_proxy_integration_test.cc

```diff
@@ -0,0 +1,223 @@
+    listener_address.reset(new Network::Address::Ipv6Instance("::1", port));
+  }
+
+  requestResponseWithListenerAddress(*listener_address);
+}
+
+// Make sure multiple clients are routed correctly to a single upstream host.
+TEST_P(UdpProxyIntegrationTest, MultipleClients) {
+  setup(1);
+  const uint32_t port = lookupPort("listener_0");
+  const auto listener_address = Network::Utility::resolveUrl(
+      fmt::format("tcp://{}:{}", Network::Test::getLoopbackAddressUrlString(version_), port));
+
+  UdpSyncClient client1(timeSystem(), version_);
+  client1.write("client1_hello", *listener_address);
+
+  UdpSyncClient client2(timeSystem(), version_);
+  client2.write("client2_hello", *listener_address);
+  client2.write("client2_hello_2", *listener_address);
+
+  Network::UdpRecvData client1_request_datagram;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForUdpDatagram(client1_request_datagram));
+  EXPECT_EQ("client1_hello", client1_request_datagram.buffer_->toString());
+
+  Network::UdpRecvData client2_request_datagram;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForUdpDatagram(client2_request_datagram));
+  EXPECT_EQ("client2_hello", client2_request_datagram.buffer_->toString());
+  ASSERT_TRUE(fake_upstreams_[0]->waitForUdpDatagram(client2_request_datagram));
+  EXPECT_EQ("client2_hello_2", client2_request_datagram.buffer_->toString());
+
+  // We should not be getting datagrams from the same peer.
+  EXPECT_NE(*client1_request_datagram.addresses_.peer_, *client2_request_datagram.addresses_.peer_);
+
+  // Send two datagrams back to client 2.
+  fake_upstreams_[0]->sendUdpDatagram("client2_world", *client2_request_datagram.addresses_.peer_);
+  fake_upstreams_[0]->sendUdpDatagram("client2_world_2",
+                                      *client2_request_datagram.addresses_.peer_);
+  Network::UdpRecvData response_datagram;
+  client2.recv(response_datagram);
+  EXPECT_EQ("client2_world", response_datagram.buffer_->toString());
+  client2.recv(response_datagram);
+  EXPECT_EQ("client2_world_2", response_datagram.buffer_->toString());
+
+  // Send 1 datagram back to client 1.
+  fake_upstreams_[0]->sendUdpDatagram("client1_world", *client1_request_datagram.addresses_.peer_);
+  client1.recv(response_datagram);
+  EXPECT_EQ("client1_world", response_datagram.buffer_->toString());
+}
+
+// Make sure sessions correctly forward to the same upstream host when there are multiple upstream
+// hosts.
+TEST_P(UdpProxyIntegrationTest, MultipleUpstreams) {
+  setup(2);
+  const uint32_t port = lookupPort("listener_0");
+  const auto listener_address = Network::Utility::resolveUrl(
+      fmt::format("tcp://{}:{}", Network::Test::getLoopbackAddressUrlString(version_), port));
+
+  UdpSyncClient client(timeSystem(), version_);
+  client.write("hello1", *listener_address);
+  client.write("hello2", *listener_address);
+  Network::UdpRecvData request_datagram;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForUdpDatagram(request_datagram));
```

Comment:
Is it certain that these packets are proxied to upstream 0?

Reply:
Yes, that is what the deterministic setting in the test setup does. I ran this test 1000 times to verify.

Reply:
I see. Thanks for pointing it out!